capsules/src/signbus/app_layer.rs

```diff
@@ -0,0 +1,127 @@
+//! Kernel implementation of signbus_app_layer
+//! apps/libsignpost/signbus_app_layer.c -> kernel/tock/capsules/src/signbus_app_layer.rs
+//! By: Justin Hsieh
+//!
+//! Usage
+//! -----
+//!
+//! ```rust
+//! // Signbus App Layer
+//! let app_layer = static_init!(
+//!     capsules::signbus::app_layer::SignbusAppLayer<'static>,
+//!     capsules::signbus::app_layer::SignbusAppLayer::new(protocol_layer,
+//!             &mut capsules::signbus::app_layer::BUFFER0,
+//!             &mut capsules::signbus::app_layer::BUFFER1
+//! ));
+//!
+//! ```
+
+
+use core::cell::Cell;
+use kernel::ReturnCode;
+use kernel::common::take_cell::TakeCell;
+
+// Capsules
+use signbus::{protocol_layer, support, test_signbus_init};
+
+/// Buffers used to concatenate message information.
+pub static mut BUFFER0: [u8; 256] = [0; 256];
+pub static mut BUFFER1: [u8; 256] = [0; 256];
+
+/// SignbusAppLayer to handle userland interaction and adding message type information.
+pub struct SignbusAppLayer<'a> {
+    protocol_layer: &'a protocol_layer::SignbusProtocolLayer<'a>,
+    payload: TakeCell<'static, [u8]>,
+    send_buf: TakeCell<'static, [u8]>,
+    client: Cell<Option<&'static test_signbus_init::SignbusInitialization<'static>>>,
+}
+
+/// AppLayerClient for I2C sending/receiving callbacks. Implemented by SignbusInitialization.
+pub trait AppLayerClient {
+    // Called when a new packet is received over I2C.
+    fn packet_received(&self, data: &'static mut [u8], length: usize, error: support::Error);
+
+    // Called when an I2C master write command is complete.
+    fn packet_sent(&self, data: &'static mut [u8], error: support::Error);
```

Comment:
Is this called on every I2C master write completion, or only when an entire packet has been sent (all fragments)? I think you mean the latter, and if so, update the comments for this and the others in this trait.